### PR TITLE
[BEAM-2049] Remove KeyedCombineFn

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

I think it is probably best to leave tweaking `StateTag` and `StateSpec` until a later PR...

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @tgroh (also any more eyes are helpful)

I'll be running all sorts of test suites offline until Jenkins is back in working order. Most are pretty easy and quick.

**Comment by kennknowles** [comment]
Confirmed the following offline, with some unrelated metrics flakes.

```
mvn clean verify -P release -pl sdks/java/core/ -am -amd
```

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/CombineFnUtilTest.java [hunk 1]
```diff
@@ -90,13 +89,12 @@ public void testToFnWithContext() throws Exception {
     }
     assertEquals(10, fnWithContext.extractOutput(accum, nullContext).intValue());
 
-    KeyedCombineFnWithContext<String, Integer, int[], Integer> keyedFnWithContext =
-        CombineFnUtil.toFnWithContext(Sum.ofIntegers().<String>asKeyedFn());
-    String key = "key";
-    accum = keyedFnWithContext.createAccumulator(key, nullContext);
+    CombineFnWithContext<Integer, int[], Integer> keyedFnWithContext =
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/util/CombineFnUtilTest.java]: not keyed

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/CombineFnUtilTest.java [hunk 2]
```diff
@@ -73,8 +72,8 @@ public void testToFnWithContextIdempotent() throws Exception {
         CombineFnUtil.toFnWithContext(Sum.ofIntegers());
     assertTrue(fnWithContext == CombineFnUtil.toFnWithContext(fnWithContext));
 
-    KeyedCombineFnWithContext<Object, Integer, int[], Integer> keyedFnWithContext =
-        CombineFnUtil.toFnWithContext(Sum.ofIntegers().asKeyedFn());
+    CombineFnWithContext<Integer, int[], Integer> keyedFnWithContext =
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/util/CombineFnUtilTest.java]: not keyed.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkKeyedCombineFn.java [hunk 3]
```diff
@@ -40,14 +40,14 @@
 
 
 /**
- * A {@link org.apache.beam.sdk.transforms.CombineFnBase.PerKeyCombineFn}
+ * A {@link org.apache.beam.sdk.transforms.CombineFnBase.GlobalCombineFn}
  * with a {@link org.apache.beam.sdk.transforms.CombineWithContext.Context} for the SparkRunner.
  */
 public class SparkKeyedCombineFn<K, InputT, AccumT, OutputT> extends SparkAbstractCombineFn {
```
Comment (tgroh) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkKeyedCombineFn.java]: This should be renamed.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/GlobalCombineFnRunners.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+import org.apache.beam.sdk.transforms.Combine.CombineFn;
+import org.apache.beam.sdk.transforms.CombineFnBase.GlobalCombineFn;
+import org.apache.beam.sdk.transforms.CombineWithContext.CombineFnWithContext;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.CombineContextFactory;
+import org.apache.beam.sdk.util.SideInputReader;
+
+/**
+ * Static utility methods that provide {@link GlobalCombineFnRunner} implementations for different
+ * keyed combine functions.
```
Comment (tgroh) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/GlobalCombineFnRunners.java]: not keyed

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkKeyedCombineFn.java (hunk 3 above)
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkKeyedCombineFn.java]: In fact the code paths are different enough that we should address this separately. It is easy, but not "just a rename", mostly due to keeping the WindowedValue wrappers in place everywhere. It is definitely a runner implementation detail that it chooses to specialize.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/CombineFnUtilTest.java (hunk 2 above)
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/util/CombineFnUtilTest.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/CombineFnUtilTest.java (hunk 1 above)
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/util/CombineFnUtilTest.java]: done

File: runners/core-java/src/main/java/org/apache/beam/runners/core/GlobalCombineFnRunners.java (hunk 4 above)
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/GlobalCombineFnRunners.java]: done

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/state/FlinkBroadcastStateInternals.java [hunk 5]
```diff
@@ -587,13 +569,13 @@ public int hashCode() {
...
     private final StateTag<? super K, CombiningState<InputT, AccumT, OutputT>> address;
-    private final Combine.KeyedCombineFn<? super K, InputT, AccumT, OutputT> combineFn;
+    private final Combine.CombineFn<InputT, AccumT, OutputT> combineFn;
     private final FlinkBroadcastStateInternals<K> flinkStateInternals;
 
     FlinkKeyedCombiningState(
         DefaultOperatorStateBackend flinkStateBackend,
         StateTag<? super K, CombiningState<InputT, AccumT, OutputT>> address,
-        Combine.KeyedCombineFn<? super K, InputT, AccumT, OutputT> combineFn,
+        Combine.CombineFn<InputT, AccumT, OutputT> combineFn,
```
Comment (kennknowles) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/state/FlinkBroadcastStateInternals.java]: Ditto here FWIW. Leaving runner-specific constructs in place.

**Comment by kennknowles** [comment]
OK. I hit a few more straggling names and tests.

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
Kicked off https://builds.apache.org/view/Beam/job/beam_PreCommit_Java_MavenInstall/10061/

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by kennknowles** [comment]
Dataflow ValidatesRunner is stalling reliably at HEAD; nothing to do with this PR. It may be infrastructure or pom or another issue. Certainly it is urgent, but I don't think this PR should wait on it. We have lots of assurances:

 - Jenkins did run WordCountIT and WindowedWordCountIT in precommit, so basic per key combines work
 - Building the new Dataflow worker involved testing akin to the Dataflow ValidatesRunner
 - I have since then manually run all of `CombineTest` on Dataflow
 - Other runners have compile time protection against this sort of change
 - I have manually run other runners' `ValidatesRunner` tests anyhow

Given the "LGTM if tests pass" I am going to merge this.

Incidentally, all of this testing I've done against the merger of HEAD and this PR also gives some assurance that our code at HEAD is OK.

**Pull request merged**